src/ledger/LedgerState.cpp

```diff
@@ -0,0 +1,801 @@
+#include "ledger/LedgerState.h"
+#include "crypto/Hex.h"
+#include "crypto/SHA.h"
+#include "ledger/LedgerEntryReference.h"
+#include "ledger/LedgerHeaderReference.h"
+#include "xdr/Stellar-ledger.h"
+#include "xdrpp/marshal.h"
+
+// TODO(jonjove): Remove this header (used for LedgerEntryKey)
+#include "ledger/EntryFrame.h"
+
+namespace stellar
+{
+using xdr::operator==;
+
+bool
+operator==(InflationVotes const& lhs, InflationVotes const& rhs)
+{
+    return (lhs.votes == rhs.votes) && (lhs.inflationDest == rhs.inflationDest);
+}
+
+LedgerStateRoot::LedgerStateRoot(Database& db) : mHasChild(false), mDb(db)
+{
+}
+
+void
+LedgerStateRoot::hasChild(bool child)
+{
+    assert(!child || !mHasChild);
+    mHasChild = child;
+}
+
+LedgerHeader const&
+LedgerStateRoot::getPreviousHeader()
+{
+    return mPreviousHeader;
+}
+
+void
+LedgerStateRoot::setPreviousHeader(LedgerHeader const& header)
+{
+    mPreviousHeader = header;
+}
+
+Database&
+LedgerStateRoot::getDatabase()
+{
+    return mDb;
+}
+
+Database::EntryCache&
+LedgerStateRoot::getCache()
+{
+    // TODO(jonjove): Do we want to continue using the cache from Database?
+    // We could refactor so the cache is a member of LedgerStateRoot
+    return mDb.getEntryCache();
+}
+
+LedgerState::LedgerState(LedgerStateRoot& root)
+    : mRoot(&root), mParent(nullptr), mChild(nullptr)
+{
+    mRoot->hasChild(true);
+}
+
+LedgerState::LedgerState(LedgerState& parent)
+    : mRoot(nullptr), mParent(&parent), mChild(nullptr)
+{
+    assert(!mParent->mChild);
+    mParent->mChild = this;
+    for (auto const& state : mParent->mState)
+    {
+        state.second->invalidate();
+    }
+    mParent->mLoadBestOfferContext.clear();
+}
+
+LedgerState::~LedgerState()
+{
+    if (mParent || mRoot)
+    {
+        rollback();
+    }
+}
+
+LedgerEntryChanges
+LedgerState::getChanges()
+{
+    assert(mParent || mRoot);
+    assert(!mChild);
+
+    LedgerEntryChanges changes;
+    for (auto const& state : mState)
+    {
+        // TODO(jonjove): Should output meta if *entry == *previous?
+        auto const& entry = state.second->ignoreInvalid().entry();
+        auto const& previous = state.second->ignoreInvalid().previousEntry();
+        if (previous)
+        {
+            changes.emplace_back(LEDGER_ENTRY_STATE);
+            changes.back().state() = *previous;
+            if (entry)
+            {
+                changes.emplace_back(LEDGER_ENTRY_UPDATED);
+                changes.back().updated() = *entry;
+            }
+            else
+            {
+                changes.emplace_back(LEDGER_ENTRY_REMOVED);
+                changes.back().removed() = state.first;
+            }
+        }
+        else
+        {
+            if (entry)
+            {
+                changes.emplace_back(LEDGER_ENTRY_CREATED);
+                changes.back().created() = *entry;
+            }
+        }
+    }
+    return changes;
+}
+
+std::vector<LedgerEntry>
+LedgerState::getLiveEntries()
+{
+    assert(mRoot);
```

Comment:
this looks like it is only allowed to be called by LedgerState just below root
what about moving that code to LedgerStateRoot?